syntax = "proto3";

package finance.options;

option java_package = "finance.options";
option java_outer_classname = "OptionPricingProto";
option go_package = "finance/options";
option csharp_namespace = "Finance.Options";

// Defines the type of option
enum OptionType {
  CALL = 0;
  PUT = 1;
}

// Input parameters for Black-Scholes option pricing
message BlackScholesParameters {
  double stock_price      = 1;  // S: Current stock price
  double strike_price     = 2;  // K: Strike price
  double risk_free_rate   = 3;  // r: Risk-free interest rate (decimal, e.g., 0.05 for 5%)
  double dividend_rate    = 4;  // q: Continuous dividend rate (decimal, e.g., 0.05 for 5%)
  double volatility       = 5;  // sigma: Volatility of the underlying asset (decimal)
  double time_to_maturity = 6;  // T: Time to maturity in years
  OptionType option_type  = 7;  // Whether this is a call or put option
}

// Request message for option pricing
message OptionPricingRequest {
  BlackScholesParameters parameters = 1;
}

// Response message containing the computed option price
message OptionPricingResponse {
  double option_price     = 1;
  double delta            = 2;
  double gamma            = 3;
  double vega             = 4;
  OptionType option_type  = 5;
}

// Service definition for option pricing
service OptionPricingService {
  // Calculates the price of an option using Black-Scholes model
  rpc CalculateOptionPrice(OptionPricingRequest) returns (OptionPricingResponse) {}
}